src/Adaptor.js

```diff
@@ -197,6 +197,90 @@ export function upsert(table, fields) {
   };
 }
 
+/**
+ * Insert or update multiple records using ON DUPLICATE KEY
+ * @public
+ * @example
+ * upsertMany(
+ *   'users', // the DB table
+ *   [
+ *     { name: 'one', email: 'one@openfn.org' },
+ *     { name: 'two', email: 'two@openfn.org' },
+ *   ]
+ * )
+ * @constructor
+ * @param {string} table - The target table
+ * @param {array} data - An array of objects or a function that returns an array
+ * @returns {Operation}
+ */
+export function upsertMany(table, data) {
+  return function (state) {
+    /*
+    Builds and exexutes a query with syntax similar to the following 
+    (see MySQL official documentation for ON DUPLICATE KEY):
+     INSERT INTO users (name, email)
+    SELECT * FROM (
+      SELECT 'one' AS name, 'one@openfn.org' AS email UNION
+      SELECT 'two' AS name, 'two@openfn.org' AS email 
+    ) AS a
+    ON DUPLICATE KEY UPDATE name = a.name, email = a.email;
+    */
+    return new Promise(function (resolve, reject) {
+      const rows = expandReferences(data)(state); 
+
+      if (!rows || rows.length === 0) {
+        console.log('No records provided; skipping upsert.');
+        resolve(state);
+      }
+
+      const columns = Object.keys(rows[0]);
+      const columnList = columns.join(', '); 
+      
+      // 1. construct the INSERT statement
+      const insertStmt = "INSERT INTO ".concat(table, " (").concat(columnList, ")"); 
+      
+      // 2. construct the subquery
+      const subqueryAlias = 'a';
+      const selectValues = rows.map(function (r) {
+        let selectColumns = columns.map(function (c) {
+          return "".concat(mysql.escape(r[c]), " as ").concat(c);
+        }).join(', ');
+        return "SELECT ".concat(selectColumns);
+      });
+      const subqueryStmt = selectValues.join(' UNION ');
```

Comment:
I'm curious about the rational behind building multiple select statements (as input rows) from data and then joining the sets using `UNION`.

The more traditional way would be to construct an known (ordered) list of keys/columns and create sets of tuples like this:

```
INSERT INTO <<table>> (col1, col2)
    VALUES
    ('val1', 24),
    ('val2', 21),
    ('val3', 29)
    AS new
ON DUPLICATE KEY UPDATE
    col1 = new.col1
```
building up individual tuples as selects seems like extra work..

Reply:
The above insert statement is not valid MySQL syntax.

The SELECT..UNION approach builds a query as recommended in MySQL documentation:

INSERT INTO t1
  SELECT * FROM (SELECT c, c+d AS e FROM t2) AS dt
  ON DUPLICATE KEY UPDATE b = e;

Reply:
I can assure you the above is valid MySQL syntax.

https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html

```
CREATE TABLE `names` (
  `id` int NOT NULL AUTO_INCREMENT,
  `first_name` varchar(100) CHARACTER SET utf8mb4 COLLATE utf8mb4_0900_ai_ci DEFAULT NULL,
  `last_name` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`)
);

INSERT INTO `names` (id, first_name, last_name)
    VALUES
    (1, 'val1', 'l1'),
    (2, 'val2', 'l2'),
    (3, 'val3', 'l3')
    AS new;
  
SELECT * FROM names;

id|first_name|last_name|
--+----------+---------+
 1|val1      |l1       |
 2|val2      |l2       |
 3|val3      |l3       |

INSERT INTO `names` (id, first_name, last_name)
    VALUES
    (1, 'foo', 'l1'),
    (2, 'val2', 'bar'),
    (3, 'val3', 'l3')
    AS new
ON DUPLICATE KEY UPDATE
    first_name = new.first_name,
    last_name = new.last_name;
    
SELECT * FROM names;

id|first_name|last_name|
--+----------+---------+
 1|foo       |l1       |
 2|val2      |bar      |
 3|val3      |l3       |
```

The documentation recommends using a derived table when inserting from other tables (with different columns), and using a `UNION` to concatenate those sets. 

The `upsertMany` and `upsert` function do not support inserting data from another table, every value is provided as data. So your sample above doesn't apply to this situation.

Reply:
This syntax works on MySQL 8, but fails when I try it on MySQL 5.6

Reply:
While the documentation refers to inserting data from other tables, I see no reason why you cannot generate the data set using a UNION of `SELECT` statements with aliases.

Reply:
```
INSERT INTO `names` (id, first_name, last_name)
    VALUES
    (1, 'foo', 'l1'),
    (2, 'val2', 'bar'),
    (3, 'val3', 'l3')
ON DUPLICATE KEY UPDATE
    first_name = VALUES(first_name),
    last_name = VALUES(last_name);
```

The MySQL 5.6 compatible version of the above example.

Reply:
Yes, that was the reason. I will used the VALUES() function, thanks!

Reply:
Ok cool, I get it now. Glad we could iron that out!

I'm sure you know that 5.6 is EOL now. But I'll try my best when moving over to `knex` to see if it's smart enough to use either aliases or `VALUES(...)` depending on the server version; and thereby keeping future versions working with your use case. :+1: